uobtheatre/payments/transaction_providers.py

```diff
@@ -0,0 +1,661 @@
+                info. When connection to a device the `device_id` should be
+                used.
+
+        Raises:
+            SquareException: If the square request returns an error
+        """
+        response = cls.client.devices.list_device_codes(
+            status=status, product_type=product_type
+        )
+        cls._handle_response_failure(response)
+
+        return response.body.get("device_codes") or []
+
+    @classmethod
+    def get_checkout(cls, checkout_id: str):
+        """Get checkout info from square for a given checkout id.
+
+        Args:
+            checkout_id (str): The id of the checkout to be fetched
+
+        Returns:
+            dict: Checkout response from square
+
+        Raises:
+            SquareException: When response is not successful
+        """
+        response = cls.client.terminal.get_terminal_checkout(checkout_id)
+        cls._handle_response_failure(response)
+
+        return response.body["checkout"]
+
+    @classmethod
+    def cancel(cls, payment: "payment_models.Transaction") -> None:
+        """Cancel terminal checkout.
+
+        Args:
+            payment (Payment): The payment to be canceled
+
+        Raises:
+            SquareException: When request is not successful
+        """
+        response = cls.client.terminal.cancel_terminal_checkout(
+            payment.provider_transaction_id
+        )
+        cls._handle_response_failure(response)
+
+    @classmethod
+    def sync_transaction(cls, payment: "payment_models.Transaction", data: dict = None):
+        """Syncs the given payment with the raw payment data"""
+        payment_id = cls.get_payment_provider_id(payment)
+
+        checkout = data if data else cls.get_checkout(payment_id)
+        payment.provider_fee = sum(
+            filter(
+                None,
+                [
+                    cls.payment_processing_fee(cls.get_payment(payment_id))
+                    for payment_id in checkout["payment_ids"]
+                ],
+            )
+        )
+        payment.status = payment_models.Transaction.Status.from_square_status(
+            checkout["status"]
+        )
+        payment.save()
+
+
+class SquareOnline(Refundable, PaymentProvider, SquarePaymentMethod):
+    """
+    Uses Square checkout api to create payment. Used for online transactions.
+
+    In order for this to work the following environment variables must be set:
+        - SQUARE_ACCESS_TOKEN
+        - SQUARE_ENVIRONMENT
+    For the sandbox values please ask the webmaster. Currently all tests that
+    require these values are marked with `square_integration`.and can therefor
+    be skipped if not required by adding `-m "not square_integration"` to the
+    test command (`make test` or `make test-v`)
+    """
+
+    description = "Square online card payment"
+
+    @classmethod
+    @property
+    def refund_providers(cls):
+        return (SquareRefund(idempotency_key=str(uuid4())),)
```

Comment:
This feels bad, is there a better way?

Reply:
Ultimately, I feel it's fine for auto refunds, we will have to generate a key regardless. For manual refund that use this method we can simply create the refund provider in the mutation as we do for payment methods. This can then use the key provided in the mtuation